File: python/partition_array_for_maximum_sum.py
#no. 1043

class Solution:
    def maxSumAfterPartitioning(self, arr: List[int], k: int) -> int:
        n = len(arr)
        dp = []
        for i in range(n):
            dp.append(0)
        dp[n - 1] = arr[n - 1]
        dp.append(0)
        for start in range(n - 2, -1, -1):
            curr_max = 0
            end = min(start + k, n)
            for i in range(start, end):
                curr_max = max(arr[i], curr_max)
                dp[start] = max(dp[start], dp[i + 1] + curr_max * (i - start + 1))
        return dp[0]

File: C++/partition_array_for_maximum_sum.cpp
// no. 1043
#include <vector>
using namespace std;

class Solution {
public:
  int maxSumAfterPartitioning(vector<int> &arr, int k) {
    int n = arr.size();
    vector<int> dp;
    for (int i = 0; i < n; i++) {
      dp.push_back(0);
    }
    dp[n - 1] = arr[n - 1];
    dp.push_back(0);
    for (int start = n - 2; start >= 0; start--) {
      int curr_max = 0;
      int end = min(start + k, n);
      for (int i = start; i < end; i++) {
        curr_max = max(arr[i], curr_max);
        dp[start] = max(dp[start], dp[i + 1] + curr_max * (i - start + 1));
      }
    }
    return dp[0];
  }
};
